wrappers/azurekeyvault/azurekeyvault.go

```diff
@@ -261,26 +272,295 @@ func (v *Wrapper) Decrypt(ctx context.Context, in *wrapping.EncryptedBlobInfo, a
+		return nil, err
+	}
+	e := new(big.Int).SetBytes(ed)
+
+	return &rsa.PublicKey{
+		N: n,
+		E: int(e.Uint64()),
+	}, nil
+}
+
+// wrapTargetKey returns wrapped key material according to the steps outlined in
+// the Azure Key Vault BYOK specification. For details, see the specification at
+// https://docs.microsoft.com/en-us/azure/key-vault/keys/byok-specification#key-transfer-blob.
+//
+// More specific details related to the wrapping and unwrapping mechanism used can be found at
+// http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cos01/pkcs11-curr-v2.40-cos01.html#_Toc408226908.
+func (v *Wrapper) wrapTargetKey(pub *rsa.PublicKey, targetKey []byte) (string, error) {
+	// Generate an ephemeral AES key
+	aesKey, err := uuid.GenerateRandomBytes(32)
+	if err != nil {
+		return "", err
+	}
+
+	// Wrap the AES key with the RSA public key using RSA-OAEP with SHA1
+	aesKeyWrapped, err := rsa.EncryptOAEP(sha1.New(), rand.Reader, pub, aesKey, []byte{})
+	if err != nil {
+		return "", err
+	}
+
+	// Wrap the target key plaintext using AES Key Wrap with Padding [RFC5649]
+	kwp, err := subtle.NewKWP(aesKey)
+	if err != nil {
+		return "", err
+	}
+	wrappedTargetKey, err := kwp.Wrap(targetKey)
+	if err != nil {
+		return "", err
+	}
+
+	// Zero the ephemeral AES key
+	for i := range aesKey {
+		aesKey[i] = 0
+	}
+
+	// Concatenate the wrapped AES key and the wrapped target
+	// key to produce the final ciphertext
+	wrappedKeyMaterial := append(aesKeyWrapped, wrappedTargetKey...)
+
+	// Return a base64 URL encoding of the ciphertext
+	return base64.RawURLEncoding.EncodeToString(wrappedKeyMaterial), nil
+}
+
+// validateKMSKey validates the given KSMKey with respect to the limitations of Azure Key Vault.
+func (v *Wrapper) validateKMSKey(key wrapping.KMSKey) error {
+	// Validate the key type and material
+	switch key.Type {
+	case wrapping.RSA2048, wrapping.RSA3072, wrapping.RSA4096:
+		if key.Material.RSAKey == nil {
+			return fmt.Errorf("must provide RSA key for key type %q", key.Type)
+		}
+	default:
+		return fmt.Errorf("%q does not support key type %q", v.Type(), key.Type)
+	}
+
+	// Validate the key purpose
+	if len(key.Purposes) == 0 {
+		return errors.New("key must have at least one purpose")
+	}
+	for _, p := range key.Purposes {
+		switch p {
+		case wrapping.Encrypt, wrapping.Decrypt, wrapping.Sign,
+			wrapping.Verify, wrapping.Wrap, wrapping.Unwrap:
+		default:
+			return fmt.Errorf("%q does not support key purpose %q", v.Type(), p)
+		}
+	}
+
+	// Validate the key protection level
+	switch key.ProtectionLevel {
+	case wrapping.HSM:
+	default:
+		return fmt.Errorf("%q does not support key protection level %q", v.Type(), key.ProtectionLevel)
+	}
+
 	return nil
 }
 
+// keyPurposesToKeyOps translates a slice of wrapping.Purpose values
+// into a slice of JSON Web Key (JWK) [RFC7517] key_ops. For details,
+// see https://tools.ietf.org/html/rfc7517#section-4.3.
+func (v *Wrapper) keyPurposesToKeyOps(purposes []wrapping.Purpose) []string {
```

Comment:
Since each implementation has it's own interpretation of what a specific `wrapping.Purpose` means, I wonder if the constants in the wrapping package should based off `iota`.

Reply:
I experimented with using `iota` for the constants and think there is benefit to keeping them as strings. One benefit is that the string values are useful in providing better error messages related to compatibility. For example, the `validateKMSKey()` function ([link](https://github.com/hashicorp/go-kms-wrapping/pull/26/files#diff-a070a70ef870193f175e6d598c8b3b9fR519)) uses the string values for error messages related to the compatibility of the `wrapping.KMSKey` and Azure. There are some other benefits, but they're mostly from the perspective of using go-kms-wrapping as an external library.

Reply:
You can have the type return a string representation of the value if you have the type satisfy the `Stringer` interface (see [crypto.Hash](https://golang.org/pkg/crypto/#Hash) as an example). I'd that say if you're using the values from these constants directly then having them as strings is fine, but if you're using them to refer by name then `iota` might be better.

Reply:
Thanks for the example! I've changed the constants to `iota` in 1e8a837. Let me know your thoughts.